### PR TITLE
MLHR-1804 MLHR-1805 MLHR-1806 #resolve #comment Refactoring of file s…

### PR DESCRIPTION
**Pull request by otterc**
…plitter and fix for kryo exception

### CODE REVIEW

**Comment by chandnisingh** [comment]
@PramodSSImmaneni  please review

File: library/src/main/java/com/datatorrent/lib/io/fs/FileSplitter.java [hunk 1]
```diff
@@ -654,6 +673,7 @@ public TimeBasedDirectoryScanner()
       discoveredFiles = new LinkedBlockingDeque<FileInfo>();
       atomicThrowable = new AtomicReference<Throwable>();
       ignoredFiles = Sets.newHashSet();
+      lock = new Lock();
```
Comment (pramodin) [inline library/src/main/java/com/datatorrent/lib/io/fs/FileSplitter.java]: Why not create this setup.. transient anyway

File: library/src/main/java/com/datatorrent/lib/io/fs/FileSplitter.java [hunk 2]
```diff
@@ -99,6 +105,8 @@
 
   protected final BasicCounters<MutableLong> fileCounters;
 
+  private final static Lock LOCK = new Lock();
```
Comment (pramodin) [inline library/src/main/java/com/datatorrent/lib/io/fs/FileSplitter.java]: Why not use Object

File: library/src/main/java/com/datatorrent/lib/io/fs/FileSplitter.java (hunk 1 above)
Comment (otterc) [inline library/src/main/java/com/datatorrent/lib/io/fs/FileSplitter.java]: If we initialize it here and don't expect the object to change, can declare it final. Don't see why it should be moved to setup()?

File: library/src/main/java/com/datatorrent/lib/io/fs/FileSplitter.java (hunk 2 above)
Comment (otterc) [inline library/src/main/java/com/datatorrent/lib/io/fs/FileSplitter.java]: Later if we want to enhance the lock then we will be able to do so. With Object we can't

**Comment by chandnisingh** [comment]
Please let me know if this can be merged?

**Comment by chandnisingh** [comment]
@PramodSSImmaneni please review and merge. This is blocking other changes.

**Comment by chandnisingh** [comment]
@PramodSSImmaneni updated the code as per review  comments. Please merge

File: library/src/main/java/com/datatorrent/lib/io/fs/FileSplitterInput.java [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+      }
+    }
+
+    /**
+     * Operations that need to be done once a scan is complete.
+     */
+    protected void scanComplete()
+    {
+      LOG.debug("scan complete {}", lastScanMillis);
+      ScannedFileInfo fileInfo = discoveredFiles.peekLast();
```
Comment (pramodin) [inline library/src/main/java/com/datatorrent/lib/io/fs/FileSplitterInput.java]: discoveredFiles could become empty by the time this executes

Comment (otterc) [inline library/src/main/java/com/datatorrent/lib/io/fs/FileSplitterInput.java]: Yes. This is tricky to fix.

**Comment by otterc** [comment]
@PramodSSImmaneni added the fix. Please review and merge

**Comment by otterc** [comment]
As per review, removed delimiter which informs the operator when a scan iteration is complete. If any specific implementation requires it, it will need to add it.

@PramodSSImmaneni  please merge the pull request

**Pull request merged**